returnn/tf/util/data.py

```diff
@@ -983,6 +983,14 @@ def declare_same_as(self, other):
     if self_derived_bases.issubset(other_derived_bases):
       # Avoid cycles on derived_from_tag. https://github.com/rwth-i6/returnn/issues/1054
       return other.declare_same_as(self)
+    if self.dyn_size is not None and other_same_base.dyn_size is not None:
```

Comment:
It's very strange that moving this up now causes different behavior, i.e. namely that some tests are passing now which were not passing before.

I now wonder whether the check here is actually not so correct, or maybe just incomplete.

The `_maybe_update` call below might update `self.dyn_size`. But `self.dyn_size` is also only set if `self.batch` is properly set. In case `self.batch is None`, which is totally valid, `self.dyn_size` is likely also not set.

What we actually want to test is: If there are any combination of `batch` and `ctx` where both `self` and `other` are defined (via `get_for_batch_ctx` logic), and those `dyn_size` is different, then we raise this exception.

I will come up with some code for this.

Reply:
Strange, my first update also did not yield the previous behavior. I am now trying again sth more.